src/Stratis.Bitcoin/P2P/NodeGroup.cs

```diff
@@ -0,0 +1,288 @@
+            this.groupSelector = groupSelector;
+            this.network = network;
+            this.parameters = parameters;
+            this.requirements = nodeRequirements;
+
+            CloneParameters();
+        }
+
+        /// <summary>The async loop we need to wait upon before we can shut down this feature.</summary>
+        private IAsyncLoop asyncLoop;
+
+        /// <summary>Factory for creating background async loop tasks.</summary>
+        private readonly IAsyncLoopFactory asyncLoopFactory;
+
+        /// <summary> True if cancellation has not been requested and the maximum number of connected nodes has not been reached.</summary>
+        private bool CanConnect
+        {
+            get { return this.disconnect.IsCancellationRequested == false && this.connectedNodes.Count < this.MaximumNodeConnections; }
+        }
+
+        private NodesCollection connectedNodes;
+        internal NodesCollection ConnectedNodes
+        {
+            get { return this.connectedNodes; }
+        }
+
+        private CancellationTokenSource disconnect;
+
+        /// <summary> How to calculate a group of an IP, by default using NBitcoin.IpExtensions.GetGroup.</summary>
+        private Func<IPEndPoint, byte[]> groupSelector;
+        private TraceCorrelation trace = new TraceCorrelation(NodeServerTrace.Trace, "Group connection");
+
+        private NodeConnectionParameters parameters;
+        internal NodeConnectionParameters Parameters
+        {
+            get { return this.parameters; }
+        }
+
+        private NodeConnectionParameters connectParameters;
+
+        internal int MaximumNodeConnections { get; set; }
+        private Network network;
+
+        internal RelatedNodeGroups RelatedGroups { get; set; }
+
+        private NodeRequirement requirements;
+        internal NodeRequirement Requirements
+        {
+            get { return this.requirements; }
+        }
+
+        internal void ConnectToPeersAsync()
+        {
+            if (this.asyncLoop != null)
+                return;
+
+            this.asyncLoop = this.asyncLoopFactory.Run(nameof(this.ConnectToPeersAsync), token =>
+            {
+                ConnectToPeers();
+                return Task.CompletedTask;
+            },
+            this.disconnect.Token,
+            repeatEvery: TimeSpans.RunOnce);
+        }
+
+        private void ConnectToPeers()
+        {
+            while (this.CanConnect)
+            {
+                Node node = null;
+
+                try
+                {
+                    var peer = SelectPeer();
+                    if (peer == null)
+                        continue;
+
+                    ConnectPeer(peer, out node);
+                }
+                catch (OperationCanceledException cancelled)
+                {
+                    if (this.disconnect.Token.IsCancellationRequested)
+                        break;
+
+                    if (node != null)
```

Comment:
Actually, yes this whole catch can be removed. The disconnect is only done IF we managed to connect but have not been able version hand shake.

We should probably introduce a proper VersionHandshake exception or somethign so that we know why the node didnt connect properly. OR return a NodeConnect result class or something that we can interrogate afterwards.